### PR TITLE
(CDAP-17332) Enabling writing bytes read/written metrics in Spark

### PR DESCRIPTION
**Pull request by chtyim**
- Sends values written to MR Counter for bytes read/written to Spark metrics
- Rewrite the OutputMetrics in Spark to ignore setting bytes written to 0
-- This is to prevent task metrics being overwritten

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicInputFormatProvider.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  private final String inputFormatClassName;
+  private final Map<String, String> configuration;
+
+  public BasicInputFormatProvider(String inputFormatClassName, Map<String, String> configuration) {
+    this.inputFormatClassName = inputFormatClassName;
+    this.configuration = ImmutableMap.copyOf(configuration);
+  }
+
+  public BasicInputFormatProvider() {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicInputFormatProvider.java:38]: why is this needed? Is it getting serialized/deserialized somewhere?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicOutputFormatProvider.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  @Override
+  public String getOutputFormatClassName() {
+    return outputFormatClassName;
+  }
+
+  @Override
+  public Map<String, String> getOutputFormatConfiguration() {
+    return outputFormatConfiguration;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicOutputFormatProvider.java:45]: should we return an unmodifiable map?

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingOutputFormat.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * @param <K> type of key to write
+ * @param <V> type of value to write
+ */
+public class TrackingOutputFormat<K, V> extends DelegatingOutputFormat<K, V> {
+
+  @Override
+  public RecordWriter<K, V> getRecordWriter(TaskAttemptContext context) throws IOException, InterruptedException {
+    OutputFormat<K, V> delegate = getDelegate(context.getConfiguration());
+
+    if (delegate instanceof FileOutputFormat) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingOutputFormat.java:41]: would be good to add comments here too on why it's not needed for files

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingInputFormat.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.io.IOException;
+
+/**
+ * An {@link InputFormat} that enables metrics tracking through {@link TaskAttemptContext} counters to Spark metrics.
+ *
+ * @param <K> type of key to read
+ * @param <V> type of value to read
+ */
+public class TrackingInputFormat<K, V> extends DelegatingInputFormat<K, V> {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingInputFormat.java:35]: It may also be useful to track how many splits an input format generates.

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/classloader/SparkClassRewriter.java [hunk 5]
```diff
@@ -224,10 +225,57 @@ public SparkClassRewriter(Function<String, InputStream> resourceLookup, boolean
...
+    if (className.equals(SPARK_OUTPUT_METRICS.getClassName())) {
+      // Rewrite the Spark OutputMetrics to skip overwriting bytes written metrics with 0.
+      return rewriteOutputMetrics(SPARK_OUTPUT_METRICS, input);
+    }
 
     return null;
   }
 
+  /**
+   * Rewrites the OutputMetrics#setBytesWritten method to skip recording zero (0) value to avoid overwritting
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/classloader/SparkClassRewriter.java]: nit: overwritting -> overwriting

Also curious, when does something call setBytesWritten(0) after setting it to a non-zero value? Is this just a Spark bug?

**Review by albertshau — APPROVED** [review]

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicInputFormatProvider.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicInputFormatProvider.java:38]: This class was extracted from an existing inner class. I see there is a Json serializer using it.

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/classloader/SparkClassRewriter.java (hunk 5 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/classloader/SparkClassRewriter.java]: Spark does it by blinding calling setBytesWritten using FileSystem statistics. It is asymmetric in Spark in how it implements bytes read (which only applies to file inputs) and bytes written (which blindly set).

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicOutputFormatProvider.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/BasicOutputFormatProvider.java:45]: good call

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingOutputFormat.java (hunk 3 above)
Comment (chtyim) [reply cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingOutputFormat.java:41]: done

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingInputFormat.java (hunk 4 above)
Comment (chtyim) [reply cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/TrackingInputFormat.java:35]: Spark doesn't have a metrics for that directly. It does track number of tasks per stage though.

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT7364

**Comment by chtyim** [comment]
Build passed.
